o1-labs/zkapp-cli
# Feature: Add compile, deploy and signining of a zkApp

**Description**
Adds the functionality to compile, deploy and sign a zkApp. In order to compile and deploy, we need to get an instance of the SmartContract class that the user wants to deploy. To get a class instance in the zkapp-cli, we look for the specified contract to deploy in the built javascript code and we expect the class to be exported somehow. 

When we try to import the user-specified class, we handle cases where the class is not exported and give an error message to the user. In the case the user-specified contract to deploy is exported, we continue on to compiling and deploying the contract.

There are some existing questions that are in the form of `TODO` comments.

Generated logged output of each artifact:
https://gist.github.com/MartinMinkov/1dd4abadb39f6ba1eb58ff6c9ee6237d

**Testing**
For anyone that wants to try this workflow locally, these are the steps.
1. Check into `feature/deploy`
2. Link the zkapp-cli using `npm link` in the root directory
3. Create a new project with `zkapp project test-deploy`
4. Create a config with `zkapp config` (use https://proxy.berkeley.minaexplorer.com/graphql for the url)
5. Request funds from faucet by using the provided link by the CLI
6. Deploy the zkapp with `zkapp deploy <network-name>`

For anybody trying this workflow above right now, you'll need to run `npm install snarkyjs` to get the latest version and then change the deploy arguments to the following in your scaffolded project:
```
deploy(args: {
    verificationKey?: string | undefined;
    zkappKey?: PrivateKey | undefined;
  })
  ```
This is fixed in #152
